cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -858,4 +861,87 @@ private ApplicationSpecificationWithChangedWorkers(ApplicationSpecification dele
+      // app not found
+      if (appSpec == null) {
+        throw new NotFoundException(programId.getParent());
+      }
+      ProgramSpecification programSpec = getExistingAppProgramSpecification(appSpec, programId);
+      // program not found
+      if (programSpec == null) {
+        throw new NotFoundException(programId);
+      }
+      return appMetadataStore.getProgramRunCount(programId);
+    }, NotFoundException.class);
+  }
+
+  @Override
+  public List<RunCountResult> getProgramRunCounts(Collection<ProgramId> programIds) {
+    return Transactionals.execute(transactional, context -> {
+      List<RunCountResult> result = new ArrayList<>();
+      AppMetadataStore appMetadataStore = getAppMetadataStore(context);
+      Map<ApplicationId, ApplicationMeta> metas =
+        appMetadataStore.getApplicationsForAppIds(
+          programIds.stream().map(ProgramId::getParent).collect(Collectors.toList()));
+
+      Set<ProgramId> existingPrograms = new HashSet<>();
+      for (ProgramId programId : programIds) {
+        ApplicationId appId = programId.getParent();
+        if (metas.containsKey(appId)) {
+          ProgramSpecification programSpec = getExistingAppProgramSpecification(metas.get(appId).getSpec(),
+                                                                                programId);
+          // program not found
+          if (programSpec == null) {
+            result.add(new RunCountResult(programId, null, new NotFoundException(programId)));
+          } else {
+            existingPrograms.add(programId);
+          }
+          // app not found
+        } else {
+          result.add(new RunCountResult(programId, null, new NotFoundException(appId)));
+        }
+      }
+
+      Map<ProgramId, Integer> runCounts = appMetadataStore.getProgramRunCounts(existingPrograms);
+      for (Map.Entry<ProgramId, Integer> entry : runCounts.entrySet()) {
+        result.add(new RunCountResult(entry.getKey(), entry.getValue(), null));
+      }
+      return result;
+    });
+  }
+
+  /**
+   * Returns the {@link ProgramSpecification} for the specified {@link ProgramId program}.
+   * @param appSpec the {@link ApplicationSpecification} of the existing application
+   * @param programId the {@link ProgramId program} for which the {@link ProgramSpecification} is requested
+   * @return the {@link ProgramSpecification} for the specified {@link ProgramId program}
+   */
+  @Nullable
+  private ProgramSpecification getExistingAppProgramSpecification(ApplicationSpecification appSpec,
+                                                                  ProgramId programId) {
+    String programName = programId.getProgram();
+    ProgramType type = programId.getType();
+    ProgramSpecification programSpec;
+    if (type == ProgramType.FLOW && appSpec.getFlows().containsKey(programName)) {
+      programSpec = appSpec.getFlows().get(programName);
```

Comment:
Use switch case and return to make it easier to read:

```java
switch (type) {
  case FLOW:
    return appSpec.getFlows().get(programName);
  case MAPREDUCE:
    return appSpec.getMapReduce().get(programName);
  case ....
  default:
    return null;
}
```

And you don't need to use `containsKey` to check as it will return `null` anyways.